steering_controllers/src/steering_controllers.cpp

```diff
@@ -0,0 +1,489 @@
+  {
+    state_interfaces_config.names.push_back(
+      params_.front_steer_names[i] + "/" + hardware_interface::HW_IF_POSITION);
+  }
+
+  return state_interfaces_config;
+}
+
+std::vector<hardware_interface::CommandInterface>
+SteeringControllers::on_export_reference_interfaces()
+{
+  reference_interfaces_.resize(nr_ref_itfs_, std::numeric_limits<double>::quiet_NaN());
+
+  std::vector<hardware_interface::CommandInterface> reference_interfaces;
+  reference_interfaces.reserve(nr_ref_itfs_);
+
+  reference_interfaces.push_back(hardware_interface::CommandInterface(
+    get_node()->get_name(), std::string("linear/") + hardware_interface::HW_IF_VELOCITY,
+    &reference_interfaces_[0]));
+
+  reference_interfaces.push_back(hardware_interface::CommandInterface(
+    get_node()->get_name(), std::string("angular/") + hardware_interface::HW_IF_POSITION,
+    &reference_interfaces_[1]));
+
+  return reference_interfaces;
+}
+
+bool SteeringControllers::on_set_chained_mode(bool chained_mode)
+{
+  // Always accept switch to/from chained mode
+  return true || chained_mode;
+}
+
+controller_interface::CallbackReturn SteeringControllers::on_activate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  // Set default value in command
+  reset_controller_reference_msg(*(input_ref_.readFromRT()), get_node());
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn SteeringControllers::on_deactivate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  // TODO(anyone): depending on number of interfaces, use definitions, e.g., `CMD_MY_ITFS`,
+  // instead of a loop
+  for (size_t i = 0; i < nr_cmd_itfs_; ++i)
+  {
+    command_interfaces_[i].set_value(std::numeric_limits<double>::quiet_NaN());
+  }
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type SteeringControllers::update_reference_from_subscribers()
+{
+  auto current_ref = *(input_ref_.readFromRT());
+  const auto age_of_last_command = get_node()->now() - (current_ref)->header.stamp;
+
+  // TODO(anyone): depending on number of interfaces, use definitions, e.g., `CMD_MY_ITFS`,
+  // instead of a loop
+  // send message only if there is no timeout
+  if (age_of_last_command <= ref_timeout_ || ref_timeout_ == rclcpp::Duration::from_seconds(0))
+  {
+    if (!std::isnan(current_ref->twist.linear.x) && !std::isnan(current_ref->twist.angular.z))
+    {
+      reference_interfaces_[0] = current_ref->twist.linear.x;
+      reference_interfaces_[1] = current_ref->twist.angular.z;
+    }
+  }
+  else
+  {
+    reference_interfaces_[0] = std::numeric_limits<double>::quiet_NaN();
+    reference_interfaces_[1] = std::numeric_limits<double>::quiet_NaN();
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+bool SteeringControllers::update_odometry(const rclcpp::Duration & period)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Odometry update not implemented");
+  return false;
+}
```

Comment:
Do we actually need this? We can make this pure-virtual class without implementation of this. So, we will not be able to compile this if there is no implementation.

This would potentially lead to restructuring of the libraries, i.e., so that we have `steering_controllers_base` and `steering_controllers` libraries (the latter is using the first one).

Any opinions about that?